### DISCUSSION TITLE
Send msg through external interface in tsn network

### DISCUSSION THREAD

**Discussion by ER404R** [omnetpp/omnetpp#1288] [Q&A]
Hey,
I will try to be as quick as possible.

We are trying to simulate a tsn network while having a standardhost with an external interface in it.

This is a short example:

NetworkConfig:
```
network ProjectTest extends TsnNetworkBase
{
    submodules:
        extNode0: StandardHost {
            @display("p=343,192");
        }
        tsnClock: TsnClock {
            @display("p=343,94");
        }
    connections:
        tsnClock.ethg++ <--> Eth100M <--> extNode0.ethg++;

}
```
For the omnet ini we are using the RealTimeScheduler, it doesnt matter whether or not we turn on or off hasTimeSync for the clock.

So in this configuration i am unable to send packets through my external interface to the simulation. The packets dont seem to reach the simulation at all. Omnet also doesnt show that any packets got dropped.

It works when i exchange the tsnClock for another tsnDevice which seems rather strange. I am at a loss here.
We are using ExtUpperEthernetInterface.

Has anyone encountered something like this before? If u need any more details about my configuration, let me know :)

**Comment by levy** [comment]
I don't know any reason why this should not work without any problem. Having time synchronization enabled or not should not have any effect on this. Try creating a simulation which can be used to reproduce the error. You can use external programs if needed to feed packets into the simulation.

**Comment by ER404R** [comment]
The simulation that i run has this error or what do u mean exactly?

My omnet.ini file is this:

```
[General]
sim-time-limit = 1s
description = "abstract"
scheduler-class = "inet::RealTimeScheduler"

*.tsnClock*.clock.oscillator.typename = "RandomDriftOscillator"
**.oscillator.changeInterval = 12.5ms
**.oscillator.driftRate = uniform(-100ppm, 100ppm)
**.oscillator.driftRateChange = uniform(-1ppm, 1ppm)
**.oscillator.driftRateChangeUpperLimit = 100ppm
**.oscillator.driftRateChangeLowerLimit = -100ppm

**.eth[*].bitrate = 100Mbps

**.mac.duplexMode = true

**.crcMode = "computed"
**.fcsMode = "computed"

**.visualizer.typename = "IntegratedMultiCanvasVisualizer"
**.visualizer.infoVsualizer[*].displayInfos = true

[Config Project_Test]
network = ProjectTest
description = ""
sim-time-limit = 3600s

*.tsnClock.hasTimeSynchronization = false
*.extNode0.eth[0].typename = "ExtLowerEthernetInterface"
*.extNode0.eth[0].device = "omnet0"
*.extNode0.eth[0].copyConfiguration = "copyFromExt"

**.referenceClock = "tsnClock.clock"

**.crcMode = "computed"
**.fcsMode = "computed"

*.tsnClock.gptp.masterPorts = ["eth0"]
*.visualizer.dataLinkVisualizer[0].displayLinks = true
*.visualizer.dataLinkVisualizer[0].activityLevel = "protocol"
*.visualizer.dataLinkVisualizer[0].packetFilter = "GptpSync"
*.visualizer.dataLinkVisualizer[0].lineColor = "blue2"

*.visualizer.numInfoVisualizers = 3
*.visualizer.infoVisualizer[0].modules = "*.tsnClock.clock"
*.tsnClock.clock.displayStringTextFormat = "time: %T"
```

This is one of the scripts i am using to send packets:
```
#!/{my python path}

from scapy.all import Ether, sendp, get_if_hwaddr, ls
from gptp.layers import PTPv2

target_mac = "0a:aa:00:00:00:02" # switch mac address

interface = "omnet0"

eth = Ether(dst=target_mac, src=get_if_hwaddr(interface))

# create sync packet
ptp_header = PTPv2(
    messageType=0,
    versionPTP=2,
    domainNumber=0,
    sequenceId=0,
    logMessageInterval=0,
    originTimestamp=0.0,
)

packet = eth / ptp_header

print(ls(packet))

sendp(packet, iface=interface)

```

I guess this should be enough to reproduce it.
I run the python script as sudo and it works when i use a StandardHost instead of the tsnClock for example.
